Add Apple Silicon bf16 Support

rahuja23 opened the pull request:
# Add Apple Silicon bf16 Support

## Overview
This PR adds support for bf16 (Brain Floating Point) operations on Apple Silicon devices, specifically targeting M1/M2 chips. The implementation includes runtime detection of bf16 support, build-time configuration, and comprehensive test coverage.

## Changes

### 1. Platform Detection
- Added `bf16_support_mac()` method to `CpuPlatform` class
- Implemented hardware capability detection for Apple Silicon
- Added PyTorch MPS backend availability check
- Added runtime verification of bf16 tensor creation

### 2. Build System Updates
- Modified `CMakeLists.txt` to support C language for bf16 testing
- Added Apple Silicon specific include paths
- Updated compiler flags for bf16 support
- Added conditional compilation for bf16 features

### 3. Test Infrastructure
- Created new test file `tests/test_apple_silicon_bf16.py`
- Added comprehensive test suite including:
  - Platform detection tests
  - bf16 support detection tests
  - PyTorch MPS backend tests
  - Tensor creation and operation tests
  - Dtype support verification



## Testing

### Unit Tests
The test suite includes the following test cases:
1. `test_platform_detection`: Verifies correct detection of Apple Silicon platform
2. `test_bf16_support_detection`: Checks bf16 support detection
3. `test_pytorch_mps_backend`: Verifies PyTorch MPS backend availability
4. `test_bf16_tensor_creation`: Tests bf16 tensor creation on MPS device
5. `test_supported_dtypes`: Verifies correct dtype support reporting

### Build Testing
- Verified build process on macOS Sonoma
- Tested with Xcode 15.4 and Apple Clang >= 15.0.0
- Confirmed successful compilation of C++ extensions

### Manual Testing
- Tested on M1 Mac
- Verified bf16 tensor operations
- Confirmed proper fallback to FP32 when bf16 is not supported

## Installation Instructions
1. Ensure macOS Sonoma or later
2. Install Xcode 15.4 or later
3. Install build dependencies:
   ```bash
   xcode-select --install
   brew install cmake ninja
   ```
4. Create and activate virtual environment:
   ```bash
   python -m venv venv
   source venv/bin/activate
   ```
5. Install vLLM with test dependencies:
   ```bash
   pip install -e .
   pip install -r requirements-test.txt
   ```

## Performance Considerations
- bf16 support provides memory efficiency benefits
- Automatic fallback to FP32 when bf16 is not supported
- No performance regression on non-Apple Silicon platforms

## Future Work
- Add performance benchmarks for bf16 operations
- Implement additional optimizations for Apple Silicon
- Add support for future Apple Silicon chips

## Related Issues
- Closes #[issue_number] (Add Apple Silicon bf16 support)

## Checklist
- [x] Added comprehensive test suite
- [x] Updated build system configuration
- [x] Added proper error handling



<!--- pyml disable-next-line no-emphasis-as-heading -->

rahuja23 committed:
Adding changes for MacM2 bf16 support

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Renaming function to check bf16 support in mac and test adaptation

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Removing dev-requirements.txt since not required

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Correction for pre-commit hook

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Formatting corrections

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

mgoin reviewed (commented):
@rahuja23 do you want to revive this PR?

rahuja23 commented on cmake/cpu_extension.cmake:
```diff
@@ -72,13 +72,32 @@ is_avx512_disabled(AVX512_DISABLED)
+        ${CMAKE_SOURCE_DIR}/cmake/test_apple_silicon_bf16.c
+        COMPILE_DEFINITIONS "-march=armv8.2-a+bf16"
+        OUTPUT_VARIABLE APPLE_SILICON_BF16_OUTPUT
+    )
+    
+    if(APPLE_SILICON_BF16_SUPPORT)
+        message(STATUS "Apple Silicon bf16 support detected")
+        add_compile_definitions(APPLE_SILICON_BF16_SUPPORT)
+        set(MARCH_FLAGS "-march=armv8.2-a+bf16+dotprod+fp16")
+    else()
+        message(STATUS "Apple Silicon bf16 support not detected")
+        set(MARCH_FLAGS "-march=armv8.2-a+dotprod+fp16")
+    endif()
+    
+    list(APPEND CXX_COMPILE_FLAGS ${MARCH_FLAGS})
 else()
     find_isa(${CPUINFO} "avx2" AVX2_FOUND)
     find_isa(${CPUINFO} "avx512f" AVX512_FOUND)
     find_isa(${CPUINFO} "POWER10" POWER10_FOUND)
     find_isa(${CPUINFO} "POWER9" POWER9_FOUND)
-    find_isa(${CPUINFO} "asimd" ASIMD_FOUND) # Check for ARM NEON support
-    find_isa(${CPUINFO} "bf16" ARM_BF16_FOUND) # Check for ARM BF16 support
+    find_isa(${CPUINFO} "asimd" ASIMD_FOUND)
+    find_isa(${CPUINFO} "bf16" ARM_BF16_FOUND)
```
Hi 
Sorry for late response. Ahh must have been removed by mistake while cleaning up my pseudo code

rahuja23 committed:
Update vllm/platforms/cpu.py

Co-authored-by: Michael Goin <mgoin64@gmail.com>
Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Adding back the comments in cpu_extension.cmake

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Fix method name consistency for bf16 support detection

- Fixed method call from bf16_support_mac() to _bf16_support_mac()
  in supported_dtypes property
- Updated test file to use correct method name with underscore prefix
- This ensures proper method visibility and call consistency

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Merge branch 'main' into feature/Macbf16Support

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
fixing pre-commit issues

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>

rahuja23 committed:
Trigger CI re-run

Update requirements/test.txt is in latest commit, triggering CI to re-run pip-compile check.

Signed-off-by: rahuja23 <51020974+rahuja23@users.noreply.github.com>
